### PR TITLE
New graph component

### PR DESCRIPTION
**Pull request by MCLiii**

### CODE REVIEW

File: Frontend/src/Components/Dashboard/Dashboard.js
Comment (jrvollmer) [inline Frontend/src/Components/Dashboard/Dashboard.js]: Why are we doing this? I thought it was for graph data. What if we instead move this API call into a child component of Dashboard that houses the left-hand side of the screen (FaultsView and the 4 panes in the lower-left) so that re-rendering quickly isn't an issue? The graphs wouldn't be directly affected by that state change since it's no longer in the parent

File: Frontend/src/Components/Dashboard/Dashboard.js
Comment (jrvollmer) [inline Frontend/src/Components/Dashboard/Dashboard.js]: Note that this would need to be removed also per previous comment

File: Frontend/package.json
Comment (jrvollmer) [inline Frontend/package.json]: I don't want to be _too_ nit-picky, but the frontend version bump from this and the data recording changes should be changed so that either:
- We wait to bump the frontend version at all until the python backend refactor is good
- We change the data recording bump to be a major one on both the backend and frontend sides, and this will be a major bump as well. Both changes implement new api calls in the backend and frontend, so it would be a non-backwards compatible change on both ends since they're versioned separately.

I think the first option will be less of a hassle. Just delete the "bump versions" commit that Jack made and remove this change from your "Getting graph to work" commit. Then, hold off on any version updates until `python-backend` is good to merge. Once it is, do a major version bump on all sections of the dashboard.

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: Move back to original position

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]:
```suggestion

// the options fed into the graph object, save regardless of datasets
```
Fixing unnecessary whitespace change

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]:
```suggestion
    maintainAspectRatio: false,
```
This is probably your issue with the graphs not fitting vertically

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: Tooltip background color?

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: Can we just remove the problematic line and keep the original formatting?

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js:202]: I could be wrong, but I have a sneaking suspicion that this could be defined within the `CustomGraph` component to help reduce the electric boogaloo of things being redefined

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js:377]: Why'd we remove this? This might be another issue causing the graph vertical overflow

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: This is for dark mode. Let's revert it

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: Revert to minimize changes

**Review by jrvollmer — CHANGES_REQUESTED** [review]
@MCLiii, let's discuss the graph changes during the work session. I also want to fix the history of the `python-backend` branch and this one, but I can take care of that. I just want to run it by you before any new changes are made here

File: package-lock.json
Comment (jrvollmer) [inline package-lock.json:1]: Shouldn't need chart dependencies in top-level package.json

File: Frontend/src/Components/Dashboard/Dashboard.js
Comment (MCLiii) [reply Frontend/src/Components/Dashboard/Dashboard.js]: Okay

File: Frontend/src/Components/Dashboard/Dashboard.js
Comment (MCLiii) [reply Frontend/src/Components/Dashboard/Dashboard.js]: 好

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (MCLiii) [reply Frontend/src/Components/Graph/CustomGraph.js]: Correct

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [reply Frontend/src/Components/Graph/CustomGraph.js]: Addressed

**Comment by jrvollmer** [comment]
@MCLiii @jrvollmer, note that using the Docker RedisTimeSeries image runs into issues with running the graphs for more than a minute. Running the module locally rather than with Docker works.

File: Backend/components/graph_api.py
Comment (jrvollmer) [inline Backend/components/graph_api.py]: We need to use min-max decimation for booleans, and we should ensure that any outstanding/non-nominal values are shown on the graph. For reference, this is a boolean value:
image

**Review by jrvollmer — CHANGES_REQUESTED** [review]
I'm currently working on some state management and styling fixes

File: Backend/components/graph_api.py
Comment (MCLiii) [reply Backend/components/graph_api.py]: Yeah, I'm thinking adding a new field to dataformat to specify the aggregation algorithm for the data

Comment (jrvollmer) [reply Backend/components/graph_api.py]: Good idea

Comment (jrvollmer) [reply Backend/components/graph_api.py]: @MCLiii, once this is added (or at least when we can use a min-max algorithm for booleans and an average for other values), the PR should pretty much be ready to go

Comment (MCLiii) [reply Backend/components/graph_api.py]: Boolean aggregation is fixed

**Review by MCLiii — COMMENTED** [review]
Need to fix tool tip not properly displaying for data with different sampling density
Screenshot 2023-12-05 at 11 08 49

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]:
```suggestion
              querylist={datasetKeys.filter(key=>!noShowKeys[key])}
              histLen={historyLength}
              colorMode={colorMode}
              optionInfo={optionInfo}
              extremes={yRange}
            />
          </Center>
```

File: package-lock.json
Comment (jrvollmer) [inline package-lock.json:1]: Need to update versions in multiple places in package-lock.json. After changing the version in `package.json`, you can run `npm i` to update the versions in `/**/package-lock.json`

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: ?

File: package.json
Comment (jrvollmer) [inline package.json:1]: A few things:
- There is no backend version update, even though the backend changed.
- The frontend version update was a major version update, so the overall version bump would be a major one as well.
- **All that said, I think that we should not update any versions for this PR since it's being added to the python-refactor branch instead of main.**

**Review by jrvollmer — CHANGES_REQUESTED** [review]
@MCLiii

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: We already registered above

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [inline Frontend/src/Components/Graph/CustomGraph.js]: Remove

File: Frontend/src/Components/Graph/CustomGraph.js
Comment (jrvollmer) [reply Frontend/src/Components/Graph/CustomGraph.js]: Reverted

File: Backend/components/graph_api.py
Comment (jrvollmer) [inline Backend/components/graph_api.py]: @MCLiii, remove or address TODO

File: Backend/core/comms.py
Comment (jrvollmer) [inline Backend/core/comms.py]: Remove print

Comment (jrvollmer) [reply Backend/core/comms.py]: done

File: Backend/components/graph_api.py
Comment (jrvollmer) [reply Backend/components/graph_api.py]: Thanks!

File: Backend/components/graph_api.py
Comment (MCLiii) [reply Backend/components/graph_api.py]: This will need to wait until signal spreadsheet is updated

Comment (jrvollmer) [reply Backend/components/graph_api.py]: Gotcha, sounds good

File: Frontend/package.json
Comment (jrvollmer) [reply Frontend/package.json]: Reverted on new-graph-component. Wait until python-backend is good to rebase onto main before moving version bump to the end of the commit history.

File: package.json
Comment (jrvollmer) [reply package.json:1]: Reverted version bump

**Review by jrvollmer — APPROVED** [review]
LGTM

**Pull request merged**
